runtime/src/main/java/org/corfudb/runtime/view/StreamsView.java

```diff
@@ -131,91 +131,75 @@ public void gc(long trimMark) {
-                        tokenResponse.getConflictKey(), tokenResponse.getConflictStream(),
-                        tokenResponse.getToken().getSequence(), abortCause,
-                        TransactionalContext.getCurrentContext());
-            }
+                // Is our token a valid type?
+                AbortCause abortCause = getAbortCauseFromToken(tokenResponse);
 
-            try {
-                if (TransactionalContext.isInTransaction()) {
-                    // If this transaction has entries that wish to capture the committed address
-                    // invoke its preCommitCallbacks with the tokenResponse from the sequencer.
-                    // Note that we might invoke the same method multiple times on retries,
-                    // which means the preCommitCallback must be idempotent.
-                    TokenResponse finalTokenResponse = tokenResponse;
-                    log.debug("append: Invoking {} preCommitListeners",
-                            TransactionalContext.getRootContext().getPreCommitListeners().size());
-                    TransactionalContext.getRootContext()
-                            .getPreCommitListeners()
-                            .forEach(e -> e.preCommitCallback(finalTokenResponse));
+                if (abortCause != null) {
+                    throw new TransactionAbortedException(
+                            conflictInfo,
+                            tokenResponse.getConflictKey(), tokenResponse.getConflictStream(),
+                            tokenResponse.getToken().getSequence(), abortCause,
+                            TransactionalContext.getCurrentContext());
                 }
 
-                // Attempt to write to the log.
-                runtime.getAddressSpaceView().write(tokenResponse, ld, cacheOption);
-                // If we're here, we succeeded, return the acquired token.
-                return tokenResponse.getSequence();
-            } catch (OverwriteException oe) {
-                // We were overwritten, get a new token and try again.
-                log.warn("append[{}]: Overwritten after {} retries, streams {}",
-                        tokenResponse.getSequence(), x,
-                        Arrays.stream(streamIDs).map(Utils::toReadableId).collect(Collectors.toSet()));
-
-                if (conflictInfo != null) {
-                    // On retry, check for conflicts only from the previous attempt position,
-                    // otherwise the transaction will always conflict with itself.
-                    conflictInfo.setSnapshotTimestamp(tokenResponse.getToken());
-                }
+                try {
+                    // Run pre-commit listeners if we are in transaction.
+                    runPreCommitListeners(tokenResponse, ld, serializeMetadata);
+                    // Attempt to write to the log.
+                    runtime.getAddressSpaceView().write(tokenResponse, ld, cacheOption);
```

Comment:
question, in AddressSpaceView.Write we call useToken which internally sets the global address to the payload, but since, we've already serialized the payload and we'll just serialize the metadata, won't this be lost in the serialized form? 

https://github.com/CorfuDB/CorfuDB/blob/4530118dc80fee2c3a896910bb4fc88ecf6775d3/runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java#L243-L245

Reply:
The address in the LogEntry is used when we sync object, for real workloads, this  block is actually redundant, because the address will be set to LogData's address when we call getPayload(), so even if we set here, it will be overwritten after getPayload() is called. However in unit test framework, on the write path, payload is not being serailized to data[], that means when getPayload() is called, it's always in object format, so it won't go to following code to set the address.
This block of code is used to workaround the unit test framework case. I moved this logic to getPayload and added comments to make it more clear.

https://github.com/CorfuDB/CorfuDB/blob/4530118dc80fee2c3a896910bb4fc88ecf6775d3/runtime/src/main/java/org/corfudb/protocols/wireprotocol/LogData.java#L94-L95

Reply:
Great, that makes it cleaner. Thanks!